codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionEncoder.java

```diff
@@ -145,6 +147,15 @@ public ChannelFuture writeHeaders(ChannelHandlerContext ctx, int streamId, Http2
         return writeHeaders(ctx, streamId, headers, 0, DEFAULT_PRIORITY_WEIGHT, false, padding, endStream, promise);
     }
 
+    private static boolean validateHeadersSentState(Http2Stream stream, Http2Headers headers, boolean isServer,
+                                                    boolean endOfStream) {
+        boolean isInformational = isServer && HttpStatusClass.valueOf(headers.status()) == INFORMATIONAL;
+        if ((isInformational || !endOfStream) && stream.isHeadersSent() || stream.isTrailersSent()) {
+            throw new IllegalStateException("Stream " + stream.id() + " sent too many headers EOS: " + endOfStream);
```

Comment:
> IIUC is generally not common practice to include throws clauses for runtime exceptions. Can you can submit a followup PR to make the changes you have in mind to clarify?

I am coming from a Guava mindset, which has a special exception for "The remote side messed up"  called VerificationException.  I think you don't depend on Guava, but it's a useful idea.  If there isn't any recovery path from such an exception, why not just make it RuntimeException, or IORuntimeException, or some such?

Followup PR for ISE is fine.  ISE is really meant for thinks like calling `read()` after `close()`.  It's a programmer error.